### DISCUSSION TITLE
Ability to use a SAML attribute as the external user ID

### DISCUSSION THREAD

**Discussion by gabriel-hoff-vectorsolutions** [zitadel/zitadel#7421] [Ideas]
After experimenting with Zitadel and [looking at the source code](https://github.com/zitadel/zitadel/blob/1d6dfadad8000161601a9456a173dbe03176ae09/internal/idp/providers/saml/session.go#L60), I realized the Subject NameID element of the SAML response is always used as the external user ID. We'd like to have the ability to configure one of the attributes present in the SAML response as the external user ID instead.

## Why is this important to us?

Some of our customers' IDPs use transient Name IDs for subjects (see `urn:oasis:names:tc:SAML:2.0:nameid-format:transient` on [section 8.3.8 of the specification](https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf)). This means these IDs cannot be relied on as they can change between logins.
At the same time, the NameID is optional according to the specification: see [section 2.4.1](https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf).

## What happens in Zitadel when transient Name IDs are used?

Zitadel fails to create/update the user with a "User already exists" error after Zitadel receives the SAML response from the IDP:

![image](https://github.com/zitadel/zitadel/assets/62257776/78f7387a-48f1-459b-9534-6240b589f026)

I believe this happens because the user's Subject Name ID differs from the previous login, while the username remains the same. Zitadel understands there is a new user logging in that has to be created, but since the username is already in use it raises that error.

## Possible Solutions

- Adding a field in the IDP configuration page/API for informing the name of an attribute whose value will be used as the external user ID
- Provide a `setExternalId` function to Actions configured to run for [external authentication flows](https://zitadel.com/docs/apis/actions/external-authentication).

## How do other solutions solve this problem?

Auth0 allows setting a User ID Attribute when configuring a SAML identity provider (more details can be found [here](https://auth0.com/docs/authenticate/identity-providers/enterprise-identity-providers/saml#set-up-mappings)).

**Comment by hifabienne** [comment]
@stebenz @livio-a @muhlemmer what do you think about this?

**Comment by muhlemmer** [comment]
It would make sense if there is an option to pick which attribute must be used as the unique user ID. I've seen this flexibility also in OIDC client libraries like symfony-oidc. Both proposed solutions sound reasonable to me.

**Reply by gabriel-hoff-vectorsolutions** [reply]
I'm happy to hear that. Should I convert this into a feature request/issue?

**Reply by hifabienne** [reply]
yes that would be great. Thanks @vs-gtadeu

**Reply by gabriel-hoff-vectorsolutions** [reply]
Here you go: https://github.com/zitadel/zitadel/issues/7483

**Reply by livio-a** [reply]
Hey @vs-gtadeu just to make sure. Your clients really use the transient format and the issue is not related to a bug, where zitadel always requested transient (https://github.com/zitadel/zitadel/pull/7745)?

We would in that case add the possibility to specifiy the attribute for the mapping directly on the IDP and not on an action.
Would that also work for you?

**Reply by gabriel-hoff-vectorsolutions** [reply]
> Your clients really use the transient format and the issue is not related to a bug, where zitadel always requested transient?

Yes, our customers' IDPs are using transient IDs, and this is unrelated to any bug in Zitadel.

> We would in that case add the possibility to specifiy the attribute for the mapping directly on the IDP and not on an action.
Would that also work for you?

Yes, that would work perfectly for us.

**Comment by fkjohn** [comment]
We've been attempting to set up SAML SSO with Entra ID, following the documentation available at  (https://zitadel.com/docs/guides/integrate/identity-providers/introduction) and have been struggling with the same issue as discussed in this discussion. 

Decoding the SAML request, it seems that Zitadel actually requests a transient Name ID:
`<samlp:AuthnRequest [Shortened]>[Shortened]<samlp:NameIDPolicy Format="**urn:oasis:names:tc:SAML:2.0:nameid-format:transient**" AllowCreate="true"/></samlp:AuthnRequest>`

When requesting the correct Name ID policy, Entra ID will respond with a persistent Name ID attribute (https://learn.microsoft.com/en-us/entra/identity-platform/single-sign-on-saml-protocol). However, the current implementation will lead to the same issue as described in this discussion. 
I don't think it is possible to adjust the Name ID Policy attribute from Zitadel at the moment, although crewjam/saml offers the configuration possibility it will default to transient. This brings me to the conclusion that the SAML integration does not work at all with SAML servers that implement the Name ID policy correctly.

Just leaving this as note for everyone that spends hours on debugging the issue, I have disclosed the details with the support.

**Reply by fforootd** [reply]
Hey, thanks for raising this.

From a first glance it looks like you are right, so let us look into it in detail.

In the meantime, I guess your customers require you to use SAML, right?
Otherwise the option to use OIDC might be a workaround.
